File: src/1_100/0039_Combination_Sum/Combination_Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates = sorted(candidates)
        res = []
        def recursive(candidates, index, cur_list, target):
            if(target == 0):
                res.append( cur_list[:] )
                return
            for i in range(index,len(candidates)):
                if(candidates[i] > target):
                    break;
                cur_list.append(candidates[i])
                recursive(candidates, i, cur_list, target-candidates[i])
                cur_list.pop()
        recursive(candidates, 0, [], target)
        return res
		
		
			
	# 优化递归代码
	def combinationSum(self, candidates, target):
		if not candidates:
			return []
		n = len(candidates)
		res = []
		candidates = sorted(candidates)
		def dfs(index,arr,tmp):
			if tmp==target:
				res.append(arr)
				return
			for i in xrange(index,n):
				val = candidates[i]
				if tmp+val>target or val>target:
					break
				dfs(i,arr+[val],tmp+val)
		dfs(0,[],0)
		return res
		
		
		
	# 另一种DFS实现
	def combinationSum(self, candidates, target):
		if not candidates:
			return []
		n = len(candidates)
		res = []
		candidates = sorted(candidates)
		def dfs(index,arr,tmp):
			if tmp==target:
				res.append(arr)
				return
			if index==n:
				return
			val = candidates[index]
			if val>target or tmp+val>target:
				return 
			dfs(index,arr+[val],tmp+val)
			dfs(index+1,arr,tmp)
		dfs(0,[],0)
		return res
		
		

File: src/1_100/0039_Combination_Sum/Combination_Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
    	Arrays.sort(candidates);
    	dfs(candidates,target,0,new Stack<Integer>());
    	return res;
    }
    
    List<List<Integer>> res = new ArrayList<List<Integer>>();
    void dfs(int[] candidates, int target, int index, Stack<Integer> cur) {
    	if(0 == target) {
    		res.add(new ArrayList<Integer>(cur));
    		return;
    	}
    	for(int i=index;i<candidates.length;i++) {
    		if(candidates[i] > target) {
    			break;
    		}
    		cur.push(candidates[i]);
    		dfs(candidates,target-candidates[i],i,cur);
    		cur.pop();
    	}
    }
}